README.md
# linearOrdering
[![Build Status](https://travis-ci.org/Yard1/linearOrdering.svg?branch=master)](https://travis-ci.org/Yard1/linearOrdering)

linearOrdering is an R package providing several methods for linear ordering of numerical data.

Currently included methods:
* Sum of ranks
* Standardized sums
* Hellwig's (Hellwig, 1968)
* TOPSIS (Hwang & Yoon, 1981).

## Installation

Use install.packages() (for the release version) or devtools (for the development version) to install linearOrdering.

```r
install.packages("linearOrdering")
```

```r
install.packages("devtools")
devtools::install_github("Yard1/linearOrdering")
```

## Usage

Documentation included in the package itself.

## Author
Antoni Baum

## References
* [Hellwig, Z. (1968). On the optimal choice of predictors. UNESCO.](https://unesdoc.unesco.org/ark:/48223/pf0000158559.locale=en)
* Yoon, K. P., & Hwang, C. L. (1981). Multiple Attribute Decision Making: Methods and Applications. New York: Springer-Verlag. ISBN: 978-3-642-48318-9

## License
[MIT](https://github.com/Yard1/linearOrdering/blob/master/LICENSE.md)


NEWS.md
# linearOrdering 1.0.1.9000

# linearOrdering 1.0.1
Fixes "data length differs from size of matrix" warnings.

# linearOrdering 1.0.0
Initial release.


R/grouping_methods.R
# Grouping methods.

#' Group linearly ordered alternatives by average.
#'
#' @param ordered_df A data frame with elements \code{alt.row}, \code{score}
#'   and \code{rank} (returned from any linear ordering method from this library).
#' @return A list including elements:
#' \describe{
#'   \item{\code{highest}}{group of highest values}
#'   \item{\code{above_average}}{group of above average values}
#'   \item{\code{below_average}}{group of below average values}
#'   \item{\code{lowest}}{group of lowest values}
#' }
#' @author Antoni Baum \email{antoni.baum@protonmail.com}
#' @examples
#' d <- matrix(rpois(12, 5), nrow = 3, ncol = 3)
#' w <- c(1, 1, 2)
#' i <- c('+', '-', '+')
#' group_by_average(sum_of_ranks(d, w, i))
#' @export
group_by_average <- function(ordered_df) {
  mean_score <- mean(ordered_df$score)
  sd_score <- sd(ordered_df$score)
  groups <-
    list(
      "highest" = c(),
      "above_average" = c(),
      "below_average" = c(),
      "lowest" = c()
    )
  for (i in 1:nrow(ordered_df))
  {
    if (ordered_df$score[i] >= mean_score + sd_score) {
      groups$highest <- c(groups$highest, ordered_df$alt.row[i])
    }
    else if (mean_score <= ordered_df$score[i] &&
        ordered_df$score[i] < mean_score + sd_score) {
      groups$above_average <-
        c(groups$above_average, ordered_df$alt.row[i])
    }
    else if (mean_score - sd_score <= ordered_df$score[i] &&
        ordered_df$score[i] < mean_score) {
      groups$below_average <-
        c(groups$below_average, ordered_df$alt.row[i])
    }
    else {
      groups$lowest <- c(groups$lowest, ordered_df$alt.row[i])
    }
  }
  return (groups)
}

#' Group linearly ordered alternatives by quartiles.
#'
#' @param ordered_df A data frame with elements \code{alt.row}, \code{score}
#'   and \code{rank} (returned from any linear ordering method from this library).
#' @return A list including elements:
#' \describe{
#'   \item{\code{first_quartile}}{group of values in first quartile}
#'   \item{\code{second_quartile}}{group of values in second quartile}
#'   \item{\code{third_quartile}}{group of values in third quartile}
#'   \item{\code{fourth_quartile}}{group of values in fourth quartile}
#' }
#' @author Antoni Baum \email{antoni.baum@protonmail.com}
#' @examples
#' d <- matrix(rpois(12, 5), nrow = 3, ncol = 3)
#' w <- c(1, 1, 2)
#' i <- c('+', '-', '+')
#' group_by_quartile(sum_of_ranks(d, w, i))
#' @export
group_by_quartile <- function(ordered_df) {
  quartiles <- quantile(ordered_df$score)
  unname(quartiles)
  groups <-
    list(
      "first_quartile" = c(),
      "second_quartile" = c(),
      "third_quartile" = c(),
      "fourth_quartile" = c()
    )
  for (i in 1:nrow(ordered_df))
  {
    if (ordered_df$score[i] <= quartiles[2]) {
      groups$first_quartile <-
        c(groups$first_quartile, ordered_df$alt.row[i])
    }
    else if (ordered_df$score[i] <= quartiles[3]) {
      groups$second_quartile <-
        c(groups$second_quartile, ordered_df$alt.row[i])
    }
    else if (ordered_df$score[i] <= quartiles[4]) {
      groups$third_quartile <-
        c(groups$third_quartile, ordered_df$alt.row[i])
    }
    else {
      groups$fourth_quartile <-
        c(groups$fourth_quartile, ordered_df$alt.row[i])
    }
  }
  return (groups)
}


R/utility_functions.R
# Utility functions

# Normalize impacts to all positive by replacing negative influences
# with their negatives.
#
# @param decision A numeric matrix with m rows for m alternatives and n columns
#   for n criterions.
# @param impacts A character vector of '+' and '-' signs for the way that each
#   criterion influences on the alternatives.
# @return A copy of \code{decision} matrix with normalized impacts.
# @author Antoni Baum \email{antoni.baum@protonmail.com}
# @examples
# d <- matrix(rpois(12, 5), nrow = 3, ncol = 3)
# i <- c('+', '-', '+')
# .normalize_impacts(d, i)
.normalize_impacts <- function(decision, impacts) {
    if (length(impacts) != ncol(decision))
        warning("length of 'impacts' is not equal to number of columns, treating every next column as positively influencing")
    decision[, impacts == "-"] <- -decision[, impacts == "-"]
    return(decision)
}

# Calculate Euclidean distance between two vectors.
#
# @param x,y Numeric vectors.
# @return The distance between the vectors.
# @author Antoni Baum \email{antoni.baum@protonmail.com}
# @examples
# x <- c(2, 5)
# y <- c(5, 7)
# .calculate_distance(x, y)
.calculate_distance <- function(x, y) {
    sqrt(sum((x - y)^2))
}

# Normalization for TOPSIS. Intended to be used with \code{apply} function
# on a target matrix.
#
# @param decision A numeric matrix with m rows for m alternatives and n columns
#   for n criterions.
# @param weights A numeric vector with length equal to number of columns in
#   decision matrix for weights of criterions.
# @return A copy of \code{decision} matrix with TOPSIS-normalized values.
# @author Antoni Baum \email{antoni.baum@protonmail.com}
# @references Hwang, C.L.; Yoon, K. (1981). Multiple Attribute Decision
# Making: Methods and Applications. New York: Springer-Verlag.
# @examples
# d <- matrix(rpois(12, 5), nrow = 3, ncol = 3)
# w <- c(1, 1, 2)
# d <- sapply(d, .topsis_normalize, w)
.topsis_normalize <- function(decision, weights) {
    weights * decision / sqrt(sum(decision^2))
}


R/linear_ordering_methods.R
# Linear ordering methods

#' Sum of ranks method.
#'
#' @param decision A numeric matrix with m rows for m alternatives and n columns
#'   for n criterions.
#' @param weights A numeric vector with length equal to number of columns in
#'   decision matrix for weights of criterions.
#' @param impacts A character vector of '+' and '-' signs for the way that each
#'   criterion influences on the alternatives.
#' @return A data frame including elements:
#' \describe{
#'   \item{\code{alt.row}}{Row number of alternatives in decision matrix.}
#'   \item{\code{score}}{Score of alternatives.}
#'   \item{\code{rank}}{Rank of alternatives based on score.}
#' }
#' Should \code{decision} be a data frame, the row names will be carried over
#' to the return value.
#' @author Antoni Baum \email{antoni.baum@protonmail.com}
#' @examples
#' d <- matrix(rpois(9, 5), nrow = 3, ncol = 3)
#' w <- c(1, 1, 2)
#' i <- c('+', '-', '+')
#' sum_of_ranks(d, w, i)
#' @export
sum_of_ranks <- function(decision, weights, impacts) {
    if (length(weights) != ncol(decision))
        stop("length of 'weights' is not equal to number of columns")
    matrix <- .normalize_impacts(decision, impacts)
    matrix <- apply(matrix, 2, rank)
    score <- as.vector(apply(matrix, 1, weighted.mean, weights))
    df <- data.frame(alt.row = 1:length(vector), score = score, rank = rank(-score), row.names = rownames(decision))
    return(df)
}

#' Standardized sums method.
#'
#' @inheritParams sum_of_ranks
#' @return A data frame including elements:
#' \describe{
#'   \item{\code{alt.row}}{Row number of alternatives in decision matrix.}
#'   \item{\code{score}}{Score of alternatives.}
#'   \item{\code{rank}}{Rank of alternatives based on score.}
#' }
#' Should \code{decision} be a data frame, the row names will be carried over
#' to the return value.
#' Should \code{decision} be a data frame, the row names will be carried over
#' to the return value.
#' @author Antoni Baum \email{antoni.baum@protonmail.com}
#' @examples
#' d <- matrix(rpois(9, 5), nrow = 3, ncol = 3)
#' w <- c(1, 1, 2)
#' i <- c('+', '-', '+')
#' standardized_sums(d, w, i)
#' @export
standardized_sums <- function(decision, weights, impacts) {
    if (length(weights) != ncol(decision))
        stop("length of 'weights' is not equal to number of columns")
    matrix <- .normalize_impacts(decision, impacts)
    matrix <- scale(matrix)
    matrix <- matrix %*% diag(weights)
    matrix <- apply(matrix, 1, mean)
    score <- as.vector(matrix - min(matrix)) / max(matrix - min(matrix))

    df <- data.frame(alt.row = 1:length(vector), score = score, rank = rank(-score), row.names = rownames(decision))
    return(df)
}

#' Hellwig's method.
#'
#' @inheritParams sum_of_ranks
#' @return A data frame including elements:
#' \describe{
#'   \item{\code{alt.row}}{Row number of alternatives in decision matrix.}
#'   \item{\code{score}}{Score of alternatives.}
#'   \item{\code{rank}}{Rank of alternatives based on score.}
#' }
#' Should \code{decision} be a data frame, the row names will be carried over
#' to the return value.
#' @author Antoni Baum \email{antoni.baum@protonmail.com}
#' @references \href{https://unesdoc.unesco.org/ark:/48223/pf0000158559.locale=en}{Hellwig, Z. (1968). On the optimal choice of predictors. UNESCO.}
#' @examples
#' d <- matrix(rpois(9, 5), nrow = 3, ncol = 3)
#' w <- c(1, 1, 2)
#' i <- c('+', '-', '+')
#' hellwig(d, w, i)
#' @export
hellwig <- function(decision, weights, impacts) {
    if (length(weights) != ncol(decision))
        stop("length of 'weights' is not equal to number of columns")
    matrix <- .normalize_impacts(decision, impacts)
    matrix <- scale(matrix)
    matrix <- matrix %*% diag(weights)
    matrix_max <- apply(matrix, 2, max)
    matrix_distance <- apply(matrix, 1, .calculate_distance, matrix_max)
    reasonable_distance <- mean(matrix_distance) + 2 * sd(matrix_distance)
    score <- as.vector(1 - matrix_distance / reasonable_distance)

    df <- data.frame(alt.row = 1:length(vector), score = score, rank = rank(-score), row.names = rownames(decision))
    return(df)
}

#' TOPSIS - the Technique for Order of Preference by Similarity to Ideal
#' Solution.
#'
#' @inheritParams sum_of_ranks
#' @return A data frame including elements:
#' \describe{
#'   \item{\code{alt.row}}{Row number of alternatives in decision matrix.}
#'   \item{\code{score}}{Score of alternatives.}
#'   \item{\code{rank}}{Rank of alternatives based on score.}
#' }
#' Should \code{decision} be a data frame, the row names will be carried over
#' to the return value.
#' @author Antoni Baum \email{antoni.baum@protonmail.com}
#' @references Yoon, K. P., & Hwang, C. L. (1981).
#' Multiple Attribute Decision Making: Methods and Applications.
#' New York: Springer-Verlag. ISBN: 978-3-642-48318-9
#' @examples
#' d <- matrix(rpois(9, 5), nrow = 3, ncol = 3)
#' w <- c(1, 1, 2)
#' i <- c('+', '-', '+')
#' topsis(d, w, i)
#' @export
topsis <- function(decision, weights, impacts) {
    if (length(weights) != ncol(decision))
        stop("length of 'weights' is not equal to number of columns")
    matrix <- .normalize_impacts(decision, impacts)
    matrix <- sapply(matrix, .topsis_normalize, weights)
    matrix_max <- apply(matrix, 2, max)
    matrix_min <- apply(matrix, 2, min)
    matrix_max_distance <- apply(matrix, 1, .calculate_distance, matrix_max)
    matrix_min_distance <- apply(matrix, 1, .calculate_distance, matrix_min)
    score <- as.vector(matrix_min_distance / (matrix_min_distance + matrix_max_distance))

    df <- data.frame(alt.row = 1:length(vector), score = score, rank = rank(-score), row.names = rownames(decision))
    return(df)
}
